File: 414-third-maximum-number/third-maximum-number.go
func thirdMax(nums []int) int {
    uniqueElemMap := make(map[int]bool)
    for _, num := range nums {
        uniqueElemMap[num] = true
    }

    uniqueElemSlice := []int{}
    for num := range uniqueElemMap {
        uniqueElemSlice = append(uniqueElemSlice, num)
    }

    sort.Slice(uniqueElemSlice, func(i, j int) bool {
        return uniqueElemSlice[i] > uniqueElemSlice[j]
    })

    if len(uniqueElemSlice) >= 3 {
        return uniqueElemSlice[2]
    } else {
        return uniqueElemSlice[0]
    }
}

File: 414-third-maximum-number/third-maximum-number.py
class Solution:
    def thirdMax(self, nums: List[int]) -> int:
        first = second = third = float('-inf')
        seen = set()
        count = 0

        for num in nums:
            if num in seen:
                continue
            
            seen.add(num)
            count += 1
            
            if num > first:
                third = second
                second = first
                first = num
            elif num > second:
                third = second
                second = num
            elif num > third:
                third = num
                
        return third if count >= 3 else first